File: java/src/LastStoneWeightII_1049.java
public class LastStoneWeightII_1049 {

    // 实际上这个问题就是问 怎么分类stones 让两个组的stone的差值最小
    public int lastStoneWeightII(int[] stones) {
        int sum = 0;
        for (int stone : stones) {
            sum += stone;
        }
        // dp[i][j] 表示前 i 个数能否形成和为 j
        boolean[][] dp = new boolean[stones.length + 1][sum / 2 + 1];
        for (int i = 0; i <= stones.length; i++) {
            dp[i][0] = true;
        }

        int maxSum = 0;
        for (int i = 1; i <= stones.length; i++) {
            for (int j = 1; j <= sum / 2; j++) {
                if (j - stones[i - 1] >= 0) {
                    dp[i][j] = dp[i - 1][j] || dp[i - 1][j - stones[i - 1]];
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
                if (dp[i][j]) maxSum = Math.max(maxSum, j);
            }
        }

        return sum - 2 * maxSum;
    }

    public static void main(String[] args) {
        System.out.println(new LastStoneWeightII_1049().lastStoneWeightII(new int[]{ 2,7,4,1,8,1 }));
    }
}

File: go/classic/lastStoneWeightII_1049.go
package classic

// 这道题题干 要求 stones 两两相撞 剩下一块儿 为剩下的石头 最小能形成的重量
// 其实就是问是否能够形成相等的两部分 因为相等的话 最后形成的石头 为 0
func lastStoneWeightII(stones []int) int {
	sum := getStonesSum(stones)
	// dp[i][j] 表示前 i 个能否形成 何为 j
	dp := make([][]bool, len(stones)+1)
	for i := 0; i < len(dp); i++ {
		dp[i] = make([]bool, sum/2+1)
		// 合为0一定可以
		dp[i][0] = true
	}
	maxSum := 0
	for i := 1; i <= len(stones); i++ {
		for j := 1; j <= sum/2; j++ {
			// 因为表示的前 i 能不能形成 j 所以 i-1 能形成的话 也是可以的
			dp[i][j] = dp[i][j] || dp[i-1][j]
			if j >= stones[i-1] {
				dp[i][j] = dp[i][j] || dp[i-1][j-stones[i-1]]
			}
			if dp[i][j] {
				maxSum = max(maxSum, j)
			}
		}
	}
	return sum - 2*maxSum
}

func getStonesSum(stones []int) (sum int) {
	for _, w := range stones {
		sum += w
	}
	return
}
